Unit tests powernets to make sure the whole station is connected.

JohnFulpWillard opened the pull request:
## About The Pull Request

I played a round on Tramstation where Medbay, Cargo, and Science were all disconnected from the powernet roundstart, and I was wandering around with a multitool figuring out why.
With modular maint rooms, I have no idea how to tell which room isn't connecting properly in maps, so now I made a unit test for it, to catch all occurrences of it, so at the very least it won't be gamebreaking.

I also fixed some problems with powernets and the_station_areas:

- GLOB.powernets was never used. Literally. The currently existing admin verb just runtimed. I just removed it in favor of SSmachines.powernets.
- GLOB.the_station_areas included Icemoon because they were never actually added to the blacklist (why isn't this just a whitelist for /area/station instead???)

I had to fix all our station-maps for this to work since they all had minor problems, here's the ones I found funny, from Kilo and Box
![image](https://user-images.githubusercontent.com/53777086/192622868-0462645c-1c9e-4262-9fd3-c656e85b7f42.png)
![image](https://user-images.githubusercontent.com/53777086/192622888-9af73208-1164-4911-a0a1-a18243764d43.png)
![image](https://user-images.githubusercontent.com/53777086/192622899-a389bf2e-832c-4b45-ba25-a767648b2a89.png)

I was hoping to include CentCom too, but the nukie outpost has cables with ZERO apcs, SMES, or anything to generate power attached to any of it. San told me it's used purely to look at through catwalk floor. I hate mappers.

## Why It's Good For The Game

I hope roundstart cable placement will be much more sane, and mappers can avoid blame of breaking powernets on maps.

## Changelog

:cl:
fix: Tramstation's departments should all be connected to the powernet at the start of a round now.
admin: Admin's power check mapping verb now actually works.
fix: Icemoon is no longer considered part of the station.
/:cl:

JohnFulpWillard committed:
Adds more cables around tramstation so powernet is connected

Mothblocks commented:
Could we unit test for this?

JohnFulpWillard committed:
Adds a unit test to check all cables

Adds a unit test checking if all cables are connected to a powernet.

JohnFulpWillard committed:
Revert "Adds a unit test to check all cables"

This reverts commit db318a8e7cb3df8c1890e904f89d4eafba7c4f53.

Mothblocks commented:
1. You want to check and call TEST_FAIL manually, since otherwise it completely exits
2. I think without a lot of work making a "it's ok if this isn't connected" subtype, you're gonna have trouble with intentionally depowered areas like in ruins?

san7890 commented:
> 2. I think without a lot of work making a "it's ok if this isn't connected" subtype, you're gonna have trouble with intentionally depowered areas like in ruins?

We could just have it only run on only station Z-levels (like the unit test that fails if any mapping errors are present) if that's a concern.

Mothblocks commented:
I guess but you should still make a subtype for if I just wanted like a decrepit room in maint

JohnFulpWillard commented:
My unit test didn't work because cables are the powernet, even if they aren't connected to the whole station. a single cable piece is a powernet

JohnFulpWillard committed:
Revert "Revert "Adds a unit test to check all cables""

This reverts commit 6c1a36b864a617df60995295464322455bb2bdc3.

Mothblocks commented:
Lol

JohnFulpWillard committed:
re-writes the unit test

I can't test this myself so I'm pushing as a commit to see if it works or not, lol.

JohnFulpWillard committed:
lol

JohnFulpWillard committed:
Revert "Adds more cables around tramstation so powernet is connected"

This reverts commit 36ae5ca4fcba878dc469061ef96c18210c22c539.

JohnFulpWillard commented:
reverted the tram fixes to see if the unit test will do its job

JohnFulpWillard committed:
actually check if it's powered, dingus

JohnFulpWillard committed:
check for station level too, nincompoop

JohnFulpWillard committed:
i cant run unit tests rn

JohnFulpWillard committed:
WTF GLOB.powernets WAS NEVER USED

JohnFulpWillard committed:
I HATE SAN7890

Nukie base has cables everywhere but not a SINGLE machine, apparently its there for FLUFF because of CATWALK FLOORS?? COULDNT EVEN BE BOTHERED TO HAVE AN SMES OR LIKE, DO WHAT CENTCOM DOES I GUESS?W HWIEGW

JohnFulpWillard committed:
don't check for solars either cheeseface

JohnFulpWillard committed:
Update cable_powernets.dm

JohnFulpWillard committed:
take SMES connection into account

JohnFulpWillard committed:
Update cable_powernets.dm

JohnFulpWillard committed:
Fixes problems with maps & support multiple fails

JohnFulpWillard committed:
Revert "Revert "Adds more cables around tramstation so powernet is connected""

This reverts commit 3f21b550d247f6fe3431814cad4a17f390c01722.

JohnFulpWillard committed:
powernet comment

JohnFulpWillard commented:
i have way too many commits for an 80 line map edit pr

Mothblocks commented on code/modules/unit_tests/cable_powernets.dm:
```diff
@@ -0,0 +1,35 @@
+			var/area/cable_area = get_area(found_cable)
+			if(!(cable_area.type in GLOB.the_station_areas) || istype(cable_area, /area/station/solars))
+				continue
+			TEST_FAIL("CABLE: [powernets] found with no nodes connected([found_cable.x], [found_cable.y], [found_cable.z])).")
```
please remove the "CABLE: " in front of all these the test name is already included

JohnFulpWillard committed:
comment and minor change to the test

JohnFulpWillard committed:
WTF

Fixes maps to pass CI
Makes icemoon blacklisted from station areas (WHY)
Adds a space to the unit test fail message

JohnFulpWillard committed:
Fixes Icebox

JohnFulpWillard commented:
@san7890 we win

JohnFulpWillard committed:
check power requirement instead

JohnFulpWillard committed:
Update cable_powernets.dm

ORCACommander commented:
congratz, you have made a useful CI tool, that also frustrates me needlessly...

JohnFulpWillard commented:
too much swag your lord
